File: timer.rb
# timer.rb
# Кухонный таймер на Ruby с потоками и неблокирующим вводом

require 'io/console'

class Timer
  attr_reader :seconds, :remaining, :paused, :running

  def initialize(seconds)
    @seconds = seconds
    @remaining = seconds
    @paused = false
    @running = false
  end

  def start
    @running = true
    Thread.new { countdown }
    Thread.new { display }
    listen_keys
  end

  private

  def countdown
    while @running
      sleep 1
      unless @paused
        @remaining -= 1
        if @remaining <= 0
          @running = false
          beep
          puts "\n⏰ Время вышло!"
          exit 0
        end
      end
    end
  end

  def display
    while @running
      m = [@remaining, 0].max / 60
      s = [@remaining, 0].max % 60
      print "\rОсталось: #{'%02d' % m}:#{'%02d' % s}   "
      sleep 0.2
    end
  end

  def listen_keys
    while @running
      char = STDIN.getch
      case char.downcase
      when 'p'
        pause
      when 'r'
        resume
      when 'q'
        stop
        puts "\nТаймер остановлен."
        exit 0
      when '+'
        @remaining += 60
      when '-'
        @remaining -= 60 if @remaining >= 60
      end
    end
  end

  def pause
    unless @paused
      @paused = true
      puts "\n⏸  Пауза"
    end
  end

  def resume
    if @paused
      @paused = false
      puts "\n▶  Возобновлено"
    end
  end

  def stop
    @running = false
  end

  def beep
    print "\007"
  end
end

def parse_time(arg)
  if arg.include?(':')
    m, s = arg.split(':')
    m.to_i * 60 + s.to_i
  else
    arg.to_i
  end
end

def show_help
  puts "Управление: p - пауза, r - возобновить, q - выход, + - +1 мин, - - -1 мин"
end

if ARGV.length > 0
  seconds = parse_time(ARGV[0])
else
  print "Введите время (сек или MM:SS): "
  input = gets.chomp
  seconds = parse_time(input)
end

timer = Timer.new(seconds)
puts "\n⏳ Таймер запущен на #{'%02d' % (seconds/60)}:#{'%02d' % (seconds%60)}"
show_help
timer.start

File: timer.js
// timer.js
// Кухонный таймер на JavaScript (Node.js) с интерактивным управлением

const readline = require('readline');
const { exec } = require('child_process');

class Timer {
    constructor(seconds) {
        this.seconds = seconds;
        this.remaining = seconds;
        this.paused = false;
        this.running = false;
        this.interval = null;
    }

    start() {
        this.running = true;
        this.interval = setInterval(() => {
            if (!this.paused) {
                this.remaining--;
                if (this.remaining <= 0) {
                    this.stop();
                    this.beep();
                    console.log('\n⏰ Время вышло!');
                    process.exit(0);
                }
            }
        }, 1000);
        this.displayLoop();
    }

    displayLoop() {
        const display = () => {
            if (this.running) {
                const m = Math.floor(Math.max(0, this.remaining) / 60);
                const s = Math.max(0, this.remaining) % 60;
                const str = `${String(m).padStart(2, '0')}:${String(s).padStart(2, '0')}`;
                readline.cursorTo(process.stdout, 0);
                process.stdout.write(`Осталось: ${str}   `);
                setTimeout(display, 200);
            }
        };
        display();
    }

    pause() {
        if (!this.paused) {
            this.paused = true;
            console.log('\n⏸  Пауза');
        }
    }

    resume() {
        if (this.paused) {
            this.paused = false;
            console.log('\n▶  Возобновлено');
        }
    }

    stop() {
        this.running = false;
        if (this.interval) {
            clearInterval(this.interval);
            this.interval = null;
        }
    }

    beep() {
        // Системный звуковой сигнал
        if (process.platform === 'win32') {
            exec('powershell -c "[console]::beep(1000,500)"');
        } else {
            process.stdout.write('\x07');
        }
    }

    addMinute() {
        this.remaining += 60;
    }

    subtractMinute() {
        if (this.remaining >= 60) {
            this.remaining -= 60;
        }
    }
}

function parseTime(arg) {
    if (arg.includes(':')) {
        const parts = arg.split(':');
        return parseInt(parts[0]) * 60 + parseInt(parts[1]);
    } else {
        return parseInt(arg);
    }
}

function showHelp() {
    console.log('Управление: p - пауза, r - возобновить, q - выход, + - +1 мин, - - -1 мин');
}

const args = process.argv.slice(2);
let seconds;
if (args.length > 0) {
    seconds = parseTime(args[0]);
} else {
    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
    });
    rl.question('Введите время (сек или MM:SS): ', (answer) => {
        rl.close();
        seconds = parseTime(answer);
        startTimer(seconds);
    });
    // Чтобы не ждать, выходим, если ввод уже обработан
    process.stdin.on('end', () => {});
    return;
}

function startTimer(secs) {
    const timer = new Timer(secs);
    console.log(`\n⏳ Таймер запущен на ${Math.floor(secs/60)}:${String(secs%60).padStart(2,'0')}`);
    showHelp();
    timer.start();

    // Управление клавишами
    readline.emitKeypressEvents(process.stdin);
    process.stdin.setRawMode(true);
    process.stdin.on('keypress', (str, key) => {
        if (key.name === 'p' || key.name === 'P') {
            timer.pause();
        } else if (key.name === 'r' || key.name === 'R') {
            timer.resume();
        } else if (key.name === 'q' || key.name === 'Q') {
            timer.stop();
            console.log('\nТаймер остановлен.');
            process.exit(0);
        } else if (str === '+') {
            timer.addMinute();
        } else if (str === '-') {
            timer.subtractMinute();
        }
    });
}

if (seconds !== undefined) {
    startTimer(seconds);
}
